lib/signal-slot-connector/signal-slot-connector.c

```diff
@@ -0,0 +1,175 @@
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "signal-slot-connector.h"
+
+struct _SignalSlotConnector
+{
+  //TODO: multithread support
+  // map<Signal, set<SlotObject>> connections;
+  GHashTable *connections;
+};
+
+GList *
+_slot_lookup_node(GList *slot_objects, Slot slot)
+{
+  for (GList *it = slot_objects; it != NULL; it = it->next)
+    {
+      Slot s = (Slot) it->data;
+
+      if (s == slot)
+        return it;
+    }
+
+  return NULL;
+}
+
+Slot
+_slot_lookup(GList *slot_objects, Slot slot)
+{
+  GList *slot_node = _slot_lookup_node(slot_objects, slot);
+  if (!slot_node)
+    return NULL;
+
+  return (Slot) slot_node->data;
+}
+
+void
+signal_slot_connect(SignalSlotConnector *self, Signal signal, Slot slot)
+{
+  g_assert(signal != NULL);
+  g_assert(slot != NULL);
+
+  GList *slots = g_hash_table_lookup(self->connections, signal);
+
+  gboolean signal_registered = (slots != NULL);
+
+  if (_slot_lookup(slots, slot))
+    {
+      msg_debug("SignalSlotConnector::connect",
+                evt_tag_printf("already_connected",
+                               "connect(connector=%p,signal=%p,slot=%p)",
+                               self, signal, slot));
+      return;
+    }
+
+  GList *new_slots = g_list_append(slots, slot);
+
+  if (!signal_registered)
+    {
+      g_hash_table_insert(self->connections, signal, new_slots);
+    }
+
+  msg_debug("SignalSlotConnector::connect",
+            evt_tag_printf("new connection registered",
+                           "connect(connector=%p,signal=%p,slot=%p)",
+                           self, signal, slot));
+}
+
+static void
+_hash_table_replace(GHashTable *hash_table, gpointer key, gpointer new_value)
+{
+  g_hash_table_steal(hash_table, key);
+  gboolean inserted_as_new = g_hash_table_insert(hash_table, key, new_value);
+  g_assert(inserted_as_new);
+}
+
+void
+signal_slot_disconnect(SignalSlotConnector *self, Signal signal, Slot slot)
+{
+  g_assert(signal != NULL);
+  g_assert(slot != NULL);
+
+  GList *slots = g_hash_table_lookup(self->connections, signal);
+
+  if (!slots)
+    return;
+
+  GList *new_slots = g_list_remove(slots, slot);
+
+  if (new_slots != slots)
+    _hash_table_replace(self->connections, signal, new_slots);
+
+  if (!new_slots)
+    {
+      g_hash_table_remove(self->connections, signal);
+    }
```

Comment:
done
(thanks!)

Reply:
hmm... this was more complicated:
When you call g_hash_table_remove, it calls the destroyfuncs for value, which frees up the list.
So we have to
 * remove the link (instead of delete)
 * when that was the last element in the list, remove the entry from the hashtable (which then free up the list) otherwise we need to free manually the removed link
 * when head of the list changed, we need to update the hashtable entry (by replacing the pointers)